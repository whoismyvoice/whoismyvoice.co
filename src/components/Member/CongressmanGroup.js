import React from 'react';
import cx from 'classnames';
import SenateStore from '../../stores/SenateStore';
import Settings from '../../data/settings.json';

// Components
import BaseComponent from '../BaseComponent';
import MemberImg from './MemberImg';
import MemberRibbon from './MemberRibbon';
import NavButton from '../Buttons/NavButton';

// Styles
import style from './../../styles/CongressmanGroup.scss';

class CongressmanGroup extends BaseComponent {
  constructor(props) {
    super(props);
    this.state = SenateStore.getMember();
  }

  shouldComponentUpdate(nextProps, nextState) {
    if (nextState.did_search && Settings.chamber === 'house' && nextState.number_representatives === 1 || Settings.chamber === 'senate' && nextState.number_representatives > 0) {
      return true;
    } else {
      return false;
    }
  }

  render() {
    const {representatives, settings, number_representatives} = this.state;

    let wrapperClasses = 'senatorWrapper',
      arrowClasses = 'arrow';

    const { chamber } = settings ? settings : Settings;

    if (representatives) {
      wrapperClasses = cx(
        ['member-wrapper'],
        {'several': representatives.length > 1}
      );
      arrowClasses = cx(
        ['arrow'],
        {'hide': representatives.length === 1}
      );
    }

    const members = (representatives || []).map(function(item, idx) {
      return <div className="member-container" key={idx}>
        <MemberImg
          bioguide={item.bioguide_id}
          chamber={chamber}
          party={item.party}
          repNumber={number_representatives}
        />
        <MemberRibbon
          name={item.full_name}
          state={item.state}
          party={item.party}
        />
        <span className={arrowClasses} id={idx}>
          <div className="line-seperator line-seperator--small"></div>
          <NavButton
            text="What can I do?"
            id={idx}
          />
        </span>
      </div>;
    });

    return <div className={wrapperClasses}>
      {members}
    </div>;
  }
}

CongressmanGroup.propTypes = {
  chamber: React.PropTypes.string
};

export default CongressmanGroup;
