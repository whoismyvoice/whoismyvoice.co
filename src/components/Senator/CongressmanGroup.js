import React from 'react';
import cx from 'classnames';

// Components
import BaseComponent from '../BaseComponent';
import SearchAddress from '../SearchAddress';
import SenatorImg from './SenatorImg';
import MemberRibbon from './MemberRibbon';
import ArrowDown from '../ArrowDown';

// Styles
import style from './../../styles/CongressmanGroup.scss';

class CongressmanGroup extends BaseComponent {
  render() {
    let searchClasses = 'search',
        wrapperClasses = 'senatorWrapper';

    const {representatives, zip_code, state_full, chamber, numRep} = this.props;

    if (representatives) {
      wrapperClasses = cx(
        ['senatorWrapper'],
        {'several': representatives.length > 1}
      );
    }

    const members = (representatives ||Â []).map(function(item, idx) {
      return <div className="member-container" key={idx}>
        <SenatorImg
          bioguide={item.bioguide_id}
          chamber={chamber}
          repNumber={numRep}
        />

        <MemberRibbon
          name={item.full_name}
          state={item.state}
          party={item.party}
        />
      </div>;
    });
    return <div className={wrapperClasses}>
      {members}
    </div>;
  }
};

CongressmanGroup.propTypes = {
  representatives: React.PropTypes.array,
  zip_code: React.PropTypes.string
};

export default CongressmanGroup;
