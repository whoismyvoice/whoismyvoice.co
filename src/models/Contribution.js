// @flow

import PropTypes from 'prop-types';

/**
 * Defines the type alias for a `Contribution` record.
 */
export type ContributionRecord = {
  amount: number,
  legislatorId: string,
  organization: string,
};

/**
 * Defines the `PropTypes` validation structure for a `Contribution` record.
 */
export const PropType = PropTypes.shape({
  amount: PropTypes.number.isRequired,
  legislatorId: PropTypes.string.isRequired,
  organization: PropTypes.string.isRequired,
});
