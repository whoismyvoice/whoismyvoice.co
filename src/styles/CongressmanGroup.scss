@import 'variables';

.member-wrapper {
	width: 90%;
	margin: 0 auto;
  margin-top: 30px;
	position: relative;

	@media screen and (max-width: $iphone6) {
		width: 95%;
    margin-top: 20px;
	}

	@media only screen
  and (min-width: 421px)
  and (max-width: 799px) {
    width: 95%;
  }

	@media only screen
  and (min-device-width: 768px)
  and (max-device-width: 1024px)
  and (orientation: landscape)
  and (-webkit-min-device-pixel-ratio: 1) {
    width: 65%;
  }

  @media screen and (max-width: 1280px) {
    margin-top: 20px;
  }

  @media screen and (max-width: 1370px) {
    margin-top: 20px;
  }
}

.member-container {
	width: 33%;
	display: inline-block;
  height: 200px;
  position: relative;

	@media screen and (max-device-width: 362px) {
    margin-top: -10px;
  }

	@media screen and (max-width: $iphone6)
	and (-webkit-min-device-pixel-ratio: 2) {
		width: 44%;
		margin-right: 2.25%;
		margin-left: 2.25%;
    height: 160px;
	}

  @media screen and (max-width: 1370px) {
    width: 30%;
    height: 190px;
  }
}

.member-name {
  width: 100%;

  @media screen and (max-width: $iphone6) {
    padding: 5px;
  }

  h2 {
    font-size: 1.2em;
    font-size: 1.1vw;
    margin-right: 5%;

    @media screen and (max-width: $iphone6) {
      margin-right: 5%;
      font-size: 0.8em;
    }

    @media only screen
    and (min-width: 421px)
    and (max-width: 799px) {
      font-size: 9px;
    }

    @media screen and (max-width: 361px) {
      font-size: 10px;
      margin-top: 0;
    }

    @media screen and (max-device-width: 413px) {
      font-size: 10px;
    }

    @media only screen
    and (min-device-width: 768px)
    and (max-device-width: 1024px)
    and (orientation: landscape)
    and (-webkit-min-device-pixel-ratio: 1) {
      font-size: 1.1em;
    }
  }
}

.member-img {
  width: 55%;
  padding: 27.5% 0;

  @media only screen
  and (min-device-width: 768px)
  and (max-device-width: 1024px)
  and (orientation: portrait)
  and (-webkit-min-device-pixel-ratio: 1) {
    width: 40%;
    padding: 20% 0;
  }

  @media only screen
  and (min-device-width: 768px)
  and (max-device-width: 1024px)
  and (orientation: landscape)
  and (-webkit-min-device-pixel-ratio: 1) {
    width: 40%;
    padding: 20% 0;
  }

  @media screen and (max-width: $iphone6) {
    width: 75%;
    padding: 37.5% 0;
    margin-bottom: 0;
  }

  @media screen
  and (min-width: 1000px)
  and (max-width: 1199px) {
    @media screen
    and (max-height: 550px) {
      margin-bottom: -5px;
    }
  }
}
