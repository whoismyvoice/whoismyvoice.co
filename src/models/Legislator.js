// @flow

import PropTypes from 'prop-types';
import { PropType as ContributionType } from './Contribution';

// Types
import type { ContributionRecord } from './Contribution';

export type BioguideId = string;
export type FecId = string;
opaque type LegislatorChamber = 'senate' | 'house' | 'unknown';
export type LegislatorIdentifier = string;

type LegislatorBioRecord = {
  birthday: string,
  gender: string,
};

export type LegislatorChannelRecord = {
  id: string,
  type: string,
};

type LegislatorIdRecord = {
  bioguide: BioguideId,
  fec: Array<FecId>,
};

type LegislatorNameRecord = {
  first?: string,
  last?: string,
  official_full: string,
};

type LegislatorTermRecord = {
  type: string,
  start: string,
  end: string,
  state: string,
  district?: string,
  party: string,
  phone: string,
};

export type LegislatorRecord = {
  id: LegislatorIdRecord,
  identifier?: string,
  name: string | LegislatorNameRecord,
  bio: LegislatorBioRecord,
  channels: Array<LegislatorChannelRecord>,
  photoUrl?: string,
  terms: Array<LegislatorTermRecord>,
};

/**
 * Defines the `PropTypes` validation structure for a `Legislator` record.
 */
export const PropType = PropTypes.shape({
  id: PropTypes.shape({
    bioguide: PropTypes.string,
    fec: PropTypes.arrayOf(PropTypes.string),
  }),
  identifier: PropTypes.string,
  name: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.shape({
      first: PropTypes.string,
      last: PropTypes.string,
      official_full: PropTypes.string.isRequired,
    }),
  ]),
  bio: PropTypes.shape({
    birthday: PropTypes.string,
    gender: PropTypes.string,
  }),
  channels: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      type: PropTypes.string,
    })
  ),
  photoUrl: PropTypes.string,
  terms: PropTypes.arrayOf(
    PropTypes.shape({
      type: PropTypes.string.isRequired,
      start: PropTypes.string.isRequired,
      end: PropTypes.string.isRequired,
      state: PropTypes.string,
      district: PropTypes.number,
      party: PropTypes.string,
    })
  ),
});

/**
 * Remove accents and diacritic marks from the given string.
 * @param {string} str to clone without 'special' characters.
 * @returns `str` with the accents and diacritic characters replaced with their
 *    plain latin variants.
 */
function stripDiacritics(str) {
  return str.normalize('NFD').replace(/[\u0300-\u036f]/g, '');
}

/**
 * Provides some convenience methods on top of the record for a legislator.
 */
export class Legislator {
  /**
   * Get the legislator's identifier from the given `record`.
   * @param {LegislatorRecord} record that matches the `PropType` defined for
   *    `Legislator` instances.
   * @returns the identifier used for legislator records for the given `record`.
   */
  static getIdentifier(record: LegislatorRecord): LegislatorIdentifier {
    PropTypes.checkPropTypes(
      { record: PropType },
      { record },
      'record',
      'Legislator#getIdentifier'
    );
    let identifier = undefined;
    if (record.identifier !== undefined) {
      identifier = record.identifier;
    } else if (typeof record.name === 'string') {
      identifier = stripDiacritics(record.name);
    } else {
      identifier = stripDiacritics(record.name.official_full);
    }
    return identifier;
  }

  /**
   * Retrieve the first contribution out of `contributions` that matches the identifier
   * of the given legislator.
   * @param {Array<ContributionRecord>} contributions records to be searched.
   * @param {LegislatorRecord|Legislator} legislator instance or record for which contributions
   *    will be searched.
   * @returns `-1` if no records from `contributions` match `legislator`, `amount`
   *    property of the contribution record otherwise.
   */
  static getContributionAmount(
    contributions: Array<ContributionRecord>,
    legislator: LegislatorRecord | Legislator
  ): number {
    PropTypes.checkPropTypes(
      { contributions: PropTypes.arrayOf(ContributionType) },
      { contributions },
      'contributions',
      'Legislator#getContributionAmount'
    );
    const id =
      legislator instanceof Legislator
        ? legislator.identifier
        : Legislator.getIdentifier(legislator);
    const contribution = contributions.find(
      record => record.legislatorId === id
    );
    if (contribution !== undefined) {
      return contribution.amount;
    } else {
      return -1;
    }
  }

  record: LegislatorRecord;

  /**
   * Create a new Legislator instance from the data provided in the given
   * `record`.
   *
   *      {
   *        "terms": [
   *          {
   *            "type": "rep|sen",
   *            "start": "<date string>",
   *            "end": "<date string>",
   *            "state": "<US State abbreviation>",
   *            "party": "<US Political Party Name>",
   *          }
   *        ]
   *      }
   *
   * @param {LegislatorRecord} record containing data about the legislator.
   */
  constructor(record: LegislatorRecord) {
    PropTypes.checkPropTypes(
      { record: PropType },
      { record },
      'record',
      'Legislator'
    );
    this.record = record;
  }

  /**
   * Gets the biographical object for the instance.
   * @returns an object containing (at least) `birthday` and `gender` properties.
   */
  get bio(): LegislatorBioRecord {
    return this.record.bio;
  }

  /**
   * Retrieve the bioguide property out of `record.id`.
   * @returns the bioguide id assigned to this legislator.
   */
  get bioguide(): BioguideId {
    return this.id.bioguide;
  }

  /**
   * Get the chamber of the legislator.
   * @returns 'house' if the legislator's current term is in the US House of
   *    Representatives; 'senate' if the legislator's current term is in the
   *    US Senate; 'unknown' otherwise.
   */
  get chamber(): LegislatorChamber {
    if (this.isSenator) {
      return 'senate';
    } else if (this.isRepresentative) {
      return 'house';
    } else {
      return 'unknown';
    }
  }

  /**
   * Retrieve the most recent term object from the given record.
   * @returns a term record for the most recent (current) term.
   */
  get currentTerm(): LegislatorTermRecord {
    const { terms } = this;
    const [currentTerm] = terms.slice(-1);
    return currentTerm;
  }

  /**
   * Retrieve the full, official name of the Legislator.
   * @returns full, official name.
   */
  get fullName(): string {
    return this.name.official_full;
  }

  /**
   * Retrieve the abbreviated gender of this legislator.
   * @returns the `gender` recorded for the legislator.
   */
  get gender(): string {
    return this.bio.gender;
  }

  /**
   * Retrieve a pronoun appropriate to the `gender` property.
   * @returns 'Them' if `gender` is anything other than 'M' or 'F'.
   */
  get genderPronoun(): string {
    switch (this.gender) {
      case 'M':
        return 'Him';
      case 'F':
        return 'Her';
      default:
        return 'Them';
    }
  }

  /**
   * Gets the id object for the instance.
   * @returns an object containing (at least) `fec` and `bioguide` properties.
   */
  get id(): LegislatorIdRecord {
    return this.record.id;
  }

  /**
   * Uses `Legislator#getIdentifier(record)` using the record provided at
   * instance creation time.
   * @returns the identifier used for legislator records for the given `record`.
   */
  get identifier(): LegislatorIdentifier {
    return Legislator.getIdentifier(this.record);
  }

  /**
   * Whether or not the `currentTerm` reflects a US Senate term.
   * @returns `true` if the `currentTerm` exists and has a Senate type.
   */
  get isSenator(): boolean {
    return this.currentTerm !== undefined && this.currentTerm.type === 'sen';
  }

  /**
   * Whether or not the `currentTerm` reflects a US House of Representatives
   * term.
   * @returns `true` if the `currentTerm` exists and has a US House of
   *    Representatives type.
   */
  get isRepresentative(): boolean {
    return this.currentTerm !== undefined && this.currentTerm.type === 'rep';
  }

  /**
   * Gets the name object for the instance.
   * @returns an object containing (at least) the `official_full` property.
   */
  get name(): LegislatorNameRecord {
    if (typeof this.record.name === 'string') {
      return { official_full: this.record.name };
    } else {
      return this.record.name;
    }
  }

  /**
   * Retrieves the US Political Party under which this legislator was elected
   * for their current term.
   * @returns current US Political Party.
   */
  get party(): string {
    return this.currentTerm.party;
  }

  /**
   * Get the phone number associated with the current term.
   * @returns The phone number associated with the current term.
   */
  get phone(): string {
    return this.currentTerm.phone;
  }

  /**
   * Get the URL for the official bioguide photo.
   * @returns a URL on theunitedstates.io.
   */
  get photoUrl(): string {
    const bioguide = this.bioguide;
    const size = 'original';
    return `https://theunitedstates.io/images/congress/${size}/${bioguide}.jpg`;
  }

  /**
   * The US State abbreviation for the state this legislator represents.
   * @returns a US State abbreviation.
   */
  get state(): string {
    return this.currentTerm.state;
  }

  /**
   * Gets the array of terms for the instance.
   * @returns an array of term records containing (at least) one record.
   */
  get terms(): Array<LegislatorTermRecord> {
    return this.record.terms;
  }

  /**
   * Get the Twitter handle of the legislator if it is provided.
   * @returns `undefined` if it isn't provided in `record`; the Twitter handle
   *    if it can be found in `record`.
   */
  get twitter(): ?string {
    const channels = this.record.channels || [];
    const channel = channels.find(channel => channel.type === 'Twitter');
    return channel === undefined ? undefined : channel.id;
  }
}

export default Legislator;
