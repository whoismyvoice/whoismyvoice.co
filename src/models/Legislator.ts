import { Contribution } from './Contribution';
import { Record as Official } from './Official';

export type TermType = 'sen' | 'rep';
export enum Party {
  Democrat = 'Democrat',
  Republican = 'Republican',
  Independent = 'Independent',
}
export type BioguideId = string;
export type FecId = string;
export enum Chamber {
  Senate,
  House,
  Unknown,
}
export type Identifier = string;

export interface BioRecord {
  birthday: string;
  gender: string;
}

export interface ChannelRecord {
  id: string;
  type: string;
}

export interface IdRecord {
  bioguide: BioguideId;
  fec: Array<FecId>;
}

export interface NameRecord {
  first?: string;
  last?: string;
  official_full: string;
}

export interface TermRecord {
  type: TermType;
  start: string;
  end: string;
  state: string;
  district?: string;
  party: Party;
  phone: string;
}

export interface Record {
  id: IdRecord;
  identifier?: string;
  name: string | NameRecord;
  bio: BioRecord;
  channels: Array<ChannelRecord>;
  photoUrl?: string;
  terms: Array<TermRecord>;
}

/**
 * Remove accents and diacritic marks from the given string.
 * @param {string} str to clone without 'special' characters.
 * @returns `str` with the accents and diacritic characters replaced with their
 *    plain latin variants.
 */
function stripDiacritics(str: string) {
  return str.normalize('NFD').replace(/[\u0300-\u036f]/g, '');
}

/**
 * Provides some convenience methods on top of the record for a legislator.
 */
export class Legislator implements Record {
  /**
   * Holds the Record used to retrieve data for getters.
   */
  private record: Record;

  /**
   * Get the legislator's identifier from the given `record`.
   * @param {Record | Official | Legislator} record that matches the `PropType`
   *    defined for `Legislator` instances.
   * @returns the identifier used for legislator records for the given `record`.
   */
  static getIdentifier(record: Record | Official | Legislator): Identifier {
    let identifier = undefined;
    if (record instanceof Legislator) {
      identifier = record.identifier;
    } else if (typeof record.name === 'string') {
      identifier = stripDiacritics(record.name);
    } else if (typeof record.name !== 'undefined') {
      identifier = stripDiacritics(record.name.official_full);
    } else {
      identifier = record.identifier || 'unknown';
    }
    return identifier;
  }

  /**
   * Retrieve the first contribution out of `contributions` that matches the identifier
   * of the given legislator.
   * @param {Array<ContributionRecord>} contributions records to be searched.
   * @param {Record|Legislator} legislator instance or record for which contributions
   *    will be searched.
   * @returns `-1` if no records from `contributions` match `legislator`, `amount`
   *    property of the contribution record otherwise.
   */
  static getContributionAmount(
    contributions: Array<Contribution>,
    legislator: Record | Legislator
  ): number {
    const id =
      legislator instanceof Legislator
        ? legislator.identifier
        : Legislator.getIdentifier(legislator);
    const contribution = contributions.find(
      record => record.legislatorId === id
    );
    if (contribution !== undefined) {
      return contribution.amount;
    } else {
      return -1;
    }
  }

  /**
   * Create a new Legislator instance from the data provided in the given
   * `record`.
   *
   *      {
   *        "terms": [
   *          {
   *            "type": "rep|sen",
   *            "start": "<date string>",
   *            "end": "<date string>",
   *            "state": "<US State abbreviation>",
   *            "party": "<US Political Party Name>",
   *          }
   *        ]
   *      }
   *
   * @param {Record} record containing data about the legislator.
   */
  constructor(record: Record) {
    this.record = record;
  }

  /**
   * Gets the biographical object for the instance.
   * @returns an object containing (at least) `birthday` and `gender` properties.
   */
  get bio(): BioRecord {
    return this.record.bio;
  }

  /**
   * Retrieve the bioguide property out of `record.id`.
   * @returns the bioguide id assigned to this legislator.
   */
  get bioguide(): BioguideId {
    return this.id.bioguide;
  }

  /**
   * Get the chamber of the legislator.
   * @returns 'house' if the legislator's current term is in the US House of
   *    Representatives; 'senate' if the legislator's current term is in the
   *    US Senate; 'unknown' otherwise.
   */
  get chamber(): Chamber {
    if (this.isSenator) {
      return Chamber.Senate;
    } else if (this.isRepresentative) {
      return Chamber.House;
    } else {
      return Chamber.Unknown;
    }
  }

  /**
   * Retrieve the channels from by which the legislator can be reached.
   * @returns an array of channels.
   */
  get channels(): Array<ChannelRecord> {
    return this.record.channels;
  }

  /**
   * Retrieve the most recent term object from the given record.
   * @returns a term record for the most recent (current) term.
   */
  get currentTerm(): TermRecord {
    const { terms } = this;
    const [currentTerm] = terms.slice(-1);
    return currentTerm;
  }

  /**
   * Retrieve the full, official name of the Legislator.
   * @returns full, official name.
   */
  get fullName(): string {
    return this.name.official_full;
  }

  /**
   * Retrieve the abbreviated gender of this legislator.
   * @returns the `gender` recorded for the legislator.
   */
  get gender(): string {
    return this.bio.gender;
  }

  /**
   * Retrieve a pronoun appropriate to the `gender` property.
   * @returns 'Them' if `gender` is anything other than 'M' or 'F'.
   */
  get genderPronoun(): string {
    switch (this.gender) {
      case 'M':
        return 'Him';
      case 'F':
        return 'Her';
      default:
        return 'Them';
    }
  }

  /**
   * Gets the id object for the instance.
   * @returns an object containing (at least) `fec` and `bioguide` properties.
   */
  get id(): IdRecord {
    return this.record.id;
  }

  /**
   * Uses `Legislator#getIdentifier(record)` using the record provided at
   * instance creation time.
   * @returns the identifier used for legislator records for the given `record`.
   */
  get identifier(): Identifier {
    return Legislator.getIdentifier(this.record);
  }

  /**
   * Whether or not the `currentTerm` reflects a US Senate term.
   * @returns `true` if the `currentTerm` exists and has a Senate type.
   */
  get isSenator(): boolean {
    return this.currentTerm !== undefined && this.currentTerm.type === 'sen';
  }

  /**
   * Whether or not the `currentTerm` reflects a US House of Representatives
   * term.
   * @returns `true` if the `currentTerm` exists and has a US House of
   *    Representatives type.
   */
  get isRepresentative(): boolean {
    return this.currentTerm !== undefined && this.currentTerm.type === 'rep';
  }

  /**
   * Gets the name object for the instance.
   * @returns an object containing (at least) the `official_full` property.
   */
  get name(): NameRecord {
    if (typeof this.record.name === 'string') {
      return { official_full: this.record.name };
    } else {
      return this.record.name;
    }
  }

  /**
   * Retrieves the US Political Party under which this legislator was elected
   * for their current term.
   * @returns current US Political Party.
   */
  get party(): Party {
    return this.currentTerm.party;
  }

  /**
   * Get the phone number associated with the current term.
   * @returns The phone number associated with the current term.
   */
  get phone(): string {
    return this.currentTerm.phone;
  }

  /**
   * Get the URL for the official bioguide photo.
   * @returns a URL on theunitedstates.io.
   */
  get photoUrl(): string {
    const bioguide = this.bioguide;
    const size = 'original';
    return `https://theunitedstates.io/images/congress/${size}/${bioguide}.jpg`;
  }

  /**
   * The US State abbreviation for the state this legislator represents.
   * @returns a US State abbreviation.
   */
  get state(): string {
    return this.currentTerm.state;
  }

  /**
   * Gets the array of terms for the instance.
   * @returns an array of term records containing (at least) one record.
   */
  get terms(): Array<TermRecord> {
    return this.record.terms;
  }

  /**
   * Get the Twitter handle of the legislator if it is provided.
   * @returns `undefined` if it isn't provided in `record`; the Twitter handle
   *    if it can be found in `record`.
   */
  get twitter(): string | undefined {
    const channels = this.record.channels || [];
    const channel = channels.find(c => c.type === 'Twitter');
    return channel === undefined ? undefined : channel.id;
  }
}

export default Legislator;
