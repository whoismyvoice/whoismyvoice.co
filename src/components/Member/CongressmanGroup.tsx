import * as React from 'react';

// Components
import MemberImg from './MemberImg';
import MemberRibbon from './MemberRibbon';
import ActionButtons from './ActionButtons';
import PaymentCounter from '../PaymentCounter';
import { Contribution, SectorContributions } from '../../models/Contribution';
import { Legislator } from '../../models/Legislator';

// Styles
import './../../styles/CongressmanGroup.scss';

interface Props {
  contributions: Array<Contribution>;
  legislators: Array<Legislator>;
  sectorContributions: SectorContributions[];
}

class CongressmanGroup extends React.Component<Props> {
  static defaultProps = {
    legislators: [],
    contributions: [],
    sectorContributions: [],
  };

  render(): JSX.Element {
    const { legislators, sectorContributions } = this.props;
    const members = legislators.map((legislator, idx) => {
      const contributionToLegislator = sectorContributions.filter(
        (group) => group.legislatorId === legislator.bioguide
      );
      const numContributors = contributionToLegislator
        .map((group) => group.contributions.length)
        .reduce((total, length) => total + length, 0);
      const totalContribution = contributionToLegislator
        .flatMap((group) =>
          group.contributions.map((contributeion) => contributeion.amount)
        )
        .reduce((total, amount) => total + amount, 0);
      return (
        <div className="member-container" key={legislator.identifier}>
          <MemberImg legislator={legislator} repNumber={idx + 1} />
          <MemberRibbon legislator={legislator} />
          <PaymentCounter
            numContributors={numContributors}
            payment={`${totalContribution}`}
          />
          <div
            className="mobile-contact-options"
            id={`legislator-contact-${idx}`}
          >
            {`Contact ${legislator.genderPronoun}`}
          </div>
          <ActionButtons legislator={legislator} />
        </div>
      );
    });

    return <div className="member-wrapper">{members}</div>;
  }
}

export default CongressmanGroup;
