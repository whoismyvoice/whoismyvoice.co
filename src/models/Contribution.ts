import { BioguideId } from './Legislator';

/**
 * Defines the type alias for a `Contribution` record.
 */
export interface Contribution {
  readonly amount: number;
  readonly legislatorId: BioguideId;
  readonly organization: string;
}

/**
 * Defines a record representing a contribution from a specified sector.
 */
export interface SectorContribution {
  readonly amount: number;
  readonly sector: string;
  readonly sectorCode: string;
}

/**
 * Associates a legislatorId with a list of `SectorContribution` records.
 */
export interface SectorContributions {
  readonly legislatorId: string;
  readonly contributions: Readonly<SectorContribution[]>;
}
