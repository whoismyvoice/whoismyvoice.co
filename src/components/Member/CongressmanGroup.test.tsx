import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import { render } from '@testing-library/react';
import CongressmanGroup from './CongressmanGroup';
import { createLegislator } from '../../models/Legislator.test';
import { Legislator } from '../../models/Legislator';

const baseProps = {
  allLegislators: [],
  legislators: [],
  sectorContributions: {},
};

it('renders without crashing', () => {
  const { container } = render(<CongressmanGroup {...baseProps} />);
  expect(container).toBeInTheDocument();
});

it('renders correctly without a legislator', () => {
  const props = {
    ...baseProps,
  };
  const { container } = render(<CongressmanGroup {...props} />);
  const wrapper = container.querySelector('.member-wrapper');
  expect(wrapper).toBeInTheDocument();
  expect(wrapper?.querySelectorAll('.member-container').length).toBe(0);
});

it('renders correctly with a legislator', () => {
  const props = {
    ...baseProps,
    legislators: [new Legislator(createLegislator('John Smith'))],
  };
  const { container, getByText } = render(<CongressmanGroup {...props} />);
  const wrapper = container.querySelector('.member-wrapper');
  expect(wrapper).toBeInTheDocument();
  expect(wrapper?.querySelectorAll('.member-container').length).toBe(
    props.legislators.length
  );
  expect(wrapper?.querySelectorAll('.member-img').length).toBe(
    props.legislators.length
  );
  expect(wrapper?.querySelectorAll('.ribbon-wrapper').length).toBe(
    props.legislators.length
  );
  expect(wrapper?.querySelectorAll('.member__payment').length).toBe(
    props.legislators.length
  );
  expect(wrapper?.querySelectorAll('.payments--graph').length).toBe(
    props.legislators.length
  );
  expect(
    getByText(`Contact ${props.legislators[0].genderPronoun}`)
  ).toBeInTheDocument();
});

it('renders correctly with a legislator and contribution', () => {
  const legislator = new Legislator(createLegislator('John Smith'));
  const props = {
    ...baseProps,
    legislators: [legislator],
    sectorContributions: {
      [legislator.bioguide]: {
        legislatorId: legislator.bioguide,
        contributions: [
          {
            amount: 1500,
            sector: 'Foo',
            sectorCode: 'F',
          },
        ],
      },
    },
  };
  const { container, getByText } = render(<CongressmanGroup {...props} />);
  const wrapper = container.querySelector('.member-wrapper');
  expect(wrapper).toBeInTheDocument();
  expect(wrapper?.querySelectorAll('.member-container').length).toBe(
    props.legislators.length
  );
  expect(wrapper?.querySelectorAll('.member-img').length).toBe(
    props.legislators.length
  );
  expect(wrapper?.querySelectorAll('.ribbon-wrapper').length).toBe(
    props.legislators.length
  );
  expect(wrapper?.querySelectorAll('.member__payment').length).toBe(
    props.legislators.length
  );
  expect(wrapper?.querySelectorAll('.payments--graph').length).toBe(
    props.legislators.length
  );
  expect(
    getByText(`Contact ${props.legislators[0].genderPronoun}`)
  ).toBeInTheDocument();
});
