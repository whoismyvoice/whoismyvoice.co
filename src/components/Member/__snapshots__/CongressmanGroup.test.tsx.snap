// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with a legislator 1`] = `
<div
  className="member-wrapper"
>
  <div
    className="member-container"
  >
    <div
      className="member-img member--grey"
    >
      <img
        alt="John Smith"
        src="https://theunitedstates.io/images/congress/original/.jpg"
      />
    </div>
    <div
      className="ribbon-wrapper"
    >
      <h1
        className="ribbon ribbon--grey"
      >
        <strong
          className="ribbon-content"
        >
          (
          I
           - 
          NY
          ) 
          John Smith
        </strong>
      </h1>
    </div>
    <span
      className="member__payment"
    >
      <span
        className="counter__accept"
      >
        Accepted
      </span>
      <span
        className="counter__digit digit__margin"
      >
        $
        <span
          className="counter__gradient-top"
        />
      </span>
      <span
        className="counter__digit"
      >
        -
        <span
          className="counter__gradient-top"
        />
      </span>
      <span
        className="counter__digit"
      >
        1
        <span
          className="counter__gradient-top"
        />
      </span>
      <span
        className="counter__from"
      >
        from 
        the NRA
      </span>
    </span>
    <div
      className="mobile-contact-options"
      id="legislator-contact-0"
    >
      Contact Them
    </div>
    <div
      className="actionButtons"
    >
      <a
        href="tel:123-123-1234"
        rel="noopener noreferrer"
        target="_blank"
      >
        <button
          className="contact__button_small undefined"
        >
          <div
            className="button__small__icon phone"
          />
          <div
            className="button__small__text"
          >
            Call Them
          </div>
        </button>
      </a>
    </div>
  </div>
</div>
`;

exports[`renders correctly with a legislator and contribution 1`] = `
<div
  className="member-wrapper"
>
  <div
    className="member-container"
  >
    <div
      className="member-img member--grey"
    >
      <img
        alt="John Smith"
        src="https://theunitedstates.io/images/congress/original/.jpg"
      />
    </div>
    <div
      className="ribbon-wrapper"
    >
      <h1
        className="ribbon ribbon--grey"
      >
        <strong
          className="ribbon-content"
        >
          (
          I
           - 
          NY
          ) 
          John Smith
        </strong>
      </h1>
    </div>
    <span
      className="member__payment"
    >
      <span
        className="counter__accept"
      >
        Accepted
      </span>
      <span
        className="counter__digit digit__margin"
      >
        $
        <span
          className="counter__gradient-top"
        />
      </span>
      <span
        className="counter__digit"
      >
        1
        <span
          className="counter__gradient-top"
        />
      </span>
      <span
        className="counter__digit"
      >
        0
        <span
          className="counter__gradient-top"
        />
      </span>
      <span
        className="counter__digit"
      >
        0
        <span
          className="counter__gradient-top"
        />
      </span>
      <span
        className="counter__digit"
      >
        0
        <span
          className="counter__gradient-top"
        />
      </span>
      <span
        className="counter__from"
      >
        from 
        the NRA
      </span>
    </span>
    <div
      className="mobile-contact-options"
      id="legislator-contact-0"
    >
      Contact Them
    </div>
    <div
      className="actionButtons"
    >
      <a
        href="tel:123-123-1234"
        rel="noopener noreferrer"
        target="_blank"
      >
        <button
          className="contact__button_small undefined"
        >
          <div
            className="button__small__icon phone"
          />
          <div
            className="button__small__text"
          >
            Call Them
          </div>
        </button>
      </a>
    </div>
  </div>
</div>
`;

exports[`renders correctly without a legislator 1`] = `
<div
  className="member-wrapper"
/>
`;
